fedimint-core/src/encoding/mod.rs

```diff
@@ -113,13 +113,74 @@ pub trait Encodable {
     }
 }
 
+/// Maximum size, in bytes, of data we are allowed to ever decode
+/// for a single value.
+pub const MAX_DECODE_SIZE: usize = 16_000_000;
+
 /// Data which can be encoded in a consensus-consistent way
 pub trait Decodable: Sized {
-    /// Decode an object with a well-defined format
+    /// Decode `Self` from a size-limited reader.
```

Comment:
I have a bad feeling about this new API, it took me a few minutes to understand how to use it, I can imagine this to be the cause for some confusion in the future, especially with the circular default impls.

At the same time I don't have a great idea hoe to make it better, so might just go with it and see how bad it actually is.

I think a trait-level comment stating "You typically implement `consensus_decode_from_finite_reader` and call `consensus_decode_from_finite_reader` in these implementations. For normal decoding outside `Decodable` implementations you likely want to use `consensus_decode`" would go a long way.

Reply:
@elsirion FWIW this exact system has been implemented in rust-bitcoin for more than a year now I think and no one complained yet. :D